src/Nethermind/Nethermind.Init/Steps/EthereumStepsManager.cs

```diff
@@ -36,153 +34,105 @@ public EthereumStepsManager(
+                if (stepBaseTypeMap.ContainsKey(stepInfo.StepBaseType))
                 {
-                    if (_logger.IsError) _logger.Error($"Unable to create instance of Ethereum runner step {stepInfo}");
-                    continue;
-                }
-
-                if (_logger.IsDebug) _logger.Debug($"Executing step: {stepInfo}");
-
-                stepInfo.Stage = StepInitializationStage.Executing;
-                startedThisRound++;
-                Task task = ExecuteStep(step, stepInfo, cancellationToken);
-
-                if (step.MustInitialize)
-                {
-                    _allPending.Enqueue(task);
+                    stepBaseTypeMap[stepInfo.StepBaseType].Add(new StepWrapper(step));
                 }
                 else
                 {
-                    stepInfo.Stage = StepInitializationStage.Complete;
+                    stepBaseTypeMap.Add(stepInfo.StepBaseType, new List<StepWrapper>() { new StepWrapper(step) });
                 }
             }
-
-            if (startedThisRound == 0 && _allPending.All(static t => t.IsCompleted))
+            List<Task> allRequiredSteps = new();
+            foreach (List<StepWrapper> steps in stepBaseTypeMap.Values)
             {
-                Interlocked.Increment(ref _foreverLoop);
-                if (_foreverLoop > 100)
+                foreach (StepWrapper stepWrapper in steps)
                 {
-                    if (_logger.IsWarn) _logger.Warn($"Didn't start any initialization steps during initialization round and all previous steps are already completed.");
+                    StepInfo stepInfo = stepInfoMap[stepWrapper.Step.GetType()];
+                    Task task = ExecuteStep(stepWrapper, stepInfo, stepBaseTypeMap, cancellationToken);
+                    if (_logger.IsDebug) _logger.Debug($"Executing step: {stepInfo}");
+
+                    if (stepWrapper.Step.MustInitialize)
+                    {
+                        allRequiredSteps.Add(task);
+                    }
                 }
             }
+            return allRequiredSteps;
         }
 
-        private async Task ExecuteStep(IStep step, StepInfo stepInfo, CancellationToken cancellationToken)
+        private async Task ExecuteStep(StepWrapper stepWrapper, StepInfo stepInfo, Dictionary<Type, List<StepWrapper>> stepBaseTypeMap, CancellationToken cancellationToken)
         {
             long startTime = Stopwatch.GetTimestamp();
             try
             {
-                await step.Execute(cancellationToken);
+                List<StepWrapper> dependencies = [];
+                foreach (Type type in stepInfo.Dependencies)
+                {
+                    if (!stepBaseTypeMap.ContainsKey(type))
+                        throw new StepDependencyException($"The dependent step {type.Name} for {stepInfo.StepType.Name} was not created.");
+                    foreach (Type dependency in stepInfo.Dependencies)
```

Comment:
Is this nested loop intended? Not just `dependencies.Add(stepBaseTypeMap[type])`?